pghoard/wal.py

```diff
@@ -42,107 +44,156 @@ class WalBlobLengthError(ValueError):
     """WAL blob is shorter than the WAL header"""
 
 
-WalHeader = namedtuple("WalHeader", ("version", "timeline", "lsn", "filename"))
+WalHeader = namedtuple("WalHeader", ("version", "lsn"))
 
 
-def read_header(blob):
-    if len(blob) < WAL_HEADER_LEN:
-        raise WalBlobLengthError(
-            "Need at least {} bytes of input to read WAL header, got {}".format(WAL_HEADER_LEN, len(blob))
+class LSN:
+
+    SEGSIZE = WAL_SEG_SIZE
+
+    def __init__(self, value, server_version: int, tli: int = None):
+        self.tli = tli
+        self.server_version = server_version
+        if isinstance(value, int):
+            self.lsn = value
+        elif isinstance(value, str):
+            log_hex, seg_hex = value.split("/", 1)
+            self.lsn = ((int(log_hex, 16) << 32) + int(seg_hex, 16))
+        else:
+            raise ValueError("LSN constructor accepts either an int, " "or a %X/%X formatted string")
+
+    @classmethod
+    def _cls_segments_per_xlogid(cls, server_version):
+        if server_version is not None and server_version < 90300:
+            return 0x0FFFFFFFF // cls.SEGSIZE
+        return 0x100000000 // cls.SEGSIZE
+
+    @property
+    def segments_per_xlogid(self):
+        return self._cls_segments_per_xlogid(self.server_version)
+
+    @classmethod
+    def from_walfile_name(cls, wal_filename, server_version):
+        n = int(wal_filename, 16)
+        tli = n >> 64
+        logid = (n >> 32) & 0xFFFFFFFF
+        segno = n & 0xFFFFFFFF
+        lsn = (logid * cls._cls_segments_per_xlogid(server_version) + segno) * cls.SEGSIZE
+        return cls(lsn, server_version, tli=tli)
+
+    @property
+    def log(self):
+        return self.lsn >> 32
+
+    @property
+    def seg(self):
```

Comment:
Any chance to give this a more self explanatory name?

Reply:
Would segno work in that case ?

Reply:
I would currently expand that in my head to sequence_number: would that be right? In any case: as someone not intimately familiar with the terminology of PG wal stuff, I would prefer spelled out names.

Reply:
Since it's an internal way of representing WAL offset, I think we should stick to the PG naming convention. In any case, those properties should be "private" as the whole point of this class is not to bother with those anymore :-)